src/main/java/com/google/opensesame/projects/ProjectData.java

```diff
@@ -10,57 +10,105 @@
+          "The project entity and its GitHub repository do not share the same repository ID.");
+    }
+
+    this.projectEntity = projectEntity;
+    this.repository = repository;
+    repositoryId = projectEntity.repositoryId;
+  }
+
+  /**
+   * Create a ProjectData object from a ProjectEntity.
+   *
+   * @param projectEntity
+   * @throws IOException Throws if there was an error retrieving the GitHub repository associated
+   *     with this ProjectEntity.
+   */
+  public ProjectData(ProjectEntity projectEntity) throws IOException {
+    this(projectEntity, GitHubGetter.getGitHub().getRepositoryById(projectEntity.repositoryId));
+  }
+
+  /**
+   * Assigns the name of this project.
+   *
+   * @return Returns the ProjectData instance with the data assigned.
+   */
+  public ProjectData withName() {
+    name = repository.getName();
+
+    return this;
+  }
+
   /**
-   * Creates project data from a ProjectEntity.
+   * Creates and assigns the data needed to preview the project on the frontend.
    *
-   * @param projectEntity A project entity from datastore or manually created.
-   * @return Returns the created ProjectData.
-   * @throws IOException Throws when there is an error communicating with the GitHub API.
+   * @return Returns the ProjectData instance with the data assigned.
+   * @throws IOException Throws if there is an error communicating with the GitHub API.
    */
-  public static ProjectData fromProjectEntity(ProjectEntity projectEntity) throws IOException {
-    GitHub gitHub = GitHubGetter.getGitHub();
-    GHRepository repository = gitHub.getRepositoryById(projectEntity.repositoryId);
+  public ProjectData withPreviewData() throws IOException {
+    this.withName();
+
+    description = repository.getDescription();
+    topicTags = repository.listTopics();
+    primaryLanguage = repository.getLanguage();
+    numMentors = projectEntity.mentorIds.size();
 
-    return fromProjectEntity(projectEntity, repository);
+    return this;
   }
 
   /**
-   * Creates project data from a ProjectEntity and the associated GitHub repository.
+   * Creates and assigns the data needed to view the full breakout page of a project on the
+   * frontend.
    *
-   * @param projectEntity A project entity from datastore or manually created.
-   * @param repository The GitHub repository associated with the project.
-   * @return Returns the created ProjectData.
-   * @throws IOException Throws when there is an error communicating with the GitHub API.
+   * @return Returns the ProjectData instance with the data assigned.
+   * @throws IOException Throws if there is an error communicating with the GitHub API.
    */
-  public static ProjectData fromProjectEntity(ProjectEntity projectEntity, GHRepository repository)
-      throws IOException {
-    ProjectPreviewData previewData =
-        ProjectPreviewData.fromProjectEntity(projectEntity, repository);
+  public ProjectData withFullData() throws IOException {
```

Comment:
Q: Having all of these constructor-helpers (withName, withPreviewData,withFullData) is definitely an efficient way to send the minimum necessary data. However, considering how large this file has now gotten, idk if this may be considered premature optimization? (honestly don't know)

Reply:
I personally don't think so. I'd be happy to hear from either Bill or Joe for their perspectives, but knowing that we are using each one of these types of data (`withName` on the mentor and dashboard pages, `withPreviewData` on the search page, and `withFullData` on the project breakout page), it doesn't seem premature.